lib/dataspec/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+use serde::{Deserialize, Serialize};
+
+use histogram::CompactHistogram;
+
+#[derive(Serialize, Deserialize)]
+pub struct Connections {
+    /// number of current connections (gauge)
+    pub current: i64,
+    /// number of total connect attempts
+    pub total: u64,
+    /// number of connections established
+    pub opened: u64,
+    /// number of connect attempts that failed
+    pub error: u64,
+    /// number of connect attempts that hit timeout
+    pub timeout: u64,
+}
+
+#[derive(Copy, Clone, Serialize, Deserialize)]
+pub struct Requests {
+    pub total: u64,
+    pub ok: u64,
+    pub reconnect: u64,
+    pub unsupported: u64,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Responses {
+    /// total number of responses
+    pub total: u64,
+    /// number of responses that were successful
+    pub ok: u64,
+    /// number of responses that were unsuccessful
+    pub error: u64,
+    /// number of responses that were missed due to timeout
+    pub timeout: u64,
+    /// number of read requests with a hit response
+    pub hit: u64,
+    /// number of read requests with a miss response
+    pub miss: u64,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ClientStats {
+    pub connections: Connections,
+    pub requests: Requests,
+    pub responses: Responses,
+    pub request_latency: CompactHistogram,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PubSub {
+    pub publishers: Publishers,
+    pub subscribers: Subscribers,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Publishers {
+    // current number of publishers
+    pub current: i64,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Subscribers {
+    // current number of subscribers
+    pub current: i64,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcPerfSnapshot {
```

Comment:
suggestion: just name this `Snapshot`? It's already in a crate called `rpcperf_dataspec` so the `RpcPerf` part can be implicit.

Reply:
I'm not attached to the name, but we already have a `Snapshot` structure here: https://github.com/iopsystems/rpc-perf/blob/main/src/metrics/mod.rs#L19

We could just end up scoping things, but I'd prefer to have explicit names. How about `OutputSnapshot` or `WindowSnapshot`? Also open to other suggestions...